src/skagent/model_analyzer.py

```diff
@@ -0,0 +1,274 @@
+"""
+
+import re
+import inspect
+from collections import defaultdict
+from skagent.model import Control, DBlock, RBlock
+
+_TOKEN_RE = re.compile(r"\b[A-Za-z_]\w*\b")
+
+
+class ModelAnalyzer:
+    """
+    Analyze a scikit-agent DBlock or RBlock and extract:
+      - node_meta: kind, agent, plate, observed for each variable
+      - edges: instant / lag / param / shock dependencies (keeps self-lag)
+      - formulas: human-readable equations for each variable
+      - plates: loop‐notation plates inferred from agents
+    """
+
+    def __init__(self, model, calibration, observables=None):
+        self.model = model
+        self.calibration = calibration.copy()
+        self.observables = set(observables or [])
+
+        # Flatten RBlock → list of DBlock(s)
+        self._blocks = []
+        self._walk_blocks()
+
+        # Storage
+        self.node_meta = {}
+        self._raw_deps = defaultdict(list)  # target → [sources…]
+        self._param_deps = defaultdict(set)  # param → {targets}
+        self._prev_deps = set()  # {(target, source), …}
+        self.edges = {"instant": [], "lag": [], "param": [], "shock": []}
+        self.formulas = {}
+        self.plates = {}
+
+    def _walk_blocks(self):
+        """Flatten RBlock into list of DBlock(s)."""
+        if isinstance(self.model, DBlock):
+            self._blocks = [self.model]
+        elif isinstance(self.model, RBlock):
+            for blk in self.model.blocks:
+                if isinstance(blk, DBlock):
+                    self._blocks.append(blk)
+                else:
+                    raise ValueError("Only one level of RBlock supported")
+        else:
+            raise ValueError("Model must be a DBlock or RBlock")
+
+    def analyze(self):
+        """Run the full analysis pipeline."""
+        self._collect_nodes()
+        self._collect_dependencies()
+
+        # Treat any referenced-but-unknown symbol as exogenous state
+        for tgt, deps in list(self._raw_deps.items()):
+            for src in deps:
+                if src not in self.node_meta:
+                    self.node_meta[src] = {
+                        "kind": "state",
+                        "agent": "global",
+                        "plate": None,
+                        "observed": False,
+                    }
+
+        self._identify_time_dependencies()
+        self._assemble_edges()
+        self._collect_formulas()
+        self._collect_plates()
+        return self
+
+    def _collect_nodes(self):
+        """Classify every variable and record its metadata."""
+        # 1) all LHS names in dynamics & reward
+        lhs = set()
+        for blk in self._blocks:
+            lhs |= set(blk.dynamics.keys())
+            lhs |= set(blk.reward.keys())
+
+        # 2) shocks
+        for blk in self._blocks:
+            for var, shock_def in blk.shocks.items():
+                agent = getattr(shock_def, "agent", "global") or "global"
```

Comment:
shocks won't have agents. the shock_def will be a Distribution of a random variable.
shocks _might_ be indexed at the block level.